### PR TITLE
Java local variable block scoping

### PR DESCRIPTION
**Pull request by smowton**
See PR https://github.com/diffblue/cbmc-testgen/pull/35 for full description.

This is the first substantial piece of the local variable scoping work, now that its pre-requisites are all merged. Note that the companion PR against cbmc-testgen is still awaiting review; making this now so this can be reviewed in parallel.

Recommend reviewing commit by commit; the first two are necessary admin, the next two make the substantial changes, and the final one cleans up code style issues.

### CODE REVIEW

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 1]
```diff
@@ -223,6 +223,49 @@ class java_bytecode_convert_methodt:public messaget
...
+  };
+
+  typedef std::map<unsigned, converted_instructiont> address_mapt;
+
+  struct block_tree_nodet
+  {
+    bool leaf;
+    std::vector<unsigned> branch_addresses;
+    std::vector<block_tree_nodet> branch;
+    block_tree_nodet() : leaf(false) {}
```
Comment (peterschrammel) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: no spaces around ":"

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 2]
```diff
@@ -495,6 +549,276 @@ static member_exprt to_member(const exprt &pointer, const exprt &fieldref)
...
+  for(;
+      checkit!=amap.end() && (checkit->first)<(findlim_block_start_address);
+      ++checkit)
+  {
+    for(auto p : checkit->second.predecessors)
+    {
+      if(p<(*findstart) || p>=findlim_block_start_address)
+      {
+        debug() << "Warning: refusing to create lexical block spanning " <<
+          (*findstart) << "-" << findlim_block_start_address <<
```
Comment (peterschrammel) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: align <<  (more instances)

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 3]
```diff
@@ -495,6 +549,276 @@ static member_exprt to_member(const exprt &pointer, const exprt &fieldref)
...
+  ++branchstart;
+  tree.branch.erase(branchstart, branchlim);
+
+  assert(tree.branch.size()==this_block_children.size());
+
+  auto branchaddriter=tree.branch_addresses.begin();
+  std::advance(branchaddriter, child_offset);
+  auto branchaddrlim=branchaddriter;
+  std::advance(branchaddrlim, nblocks);
+  newnode.branch_addresses.insert(newnode.branch_addresses.begin(),
```
Comment (peterschrammel) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: put params on next line

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 4]
```diff
@@ -1566,18 +1873,85 @@ codet java_bytecode_convert_methodt::convert_instructions(
...
+
+  // The block tree node mirrors the block structure of root_block,
+  // indexing the Java PCs were each subblock starts and ends.
+  block_tree_nodet root;
+  code_blockt root_block;
+
+  bool start_new_block=true;
+  for(auto ait=address_map.begin(), aend=address_map.end(); ait!=aend; ++ait)
   {
+    const auto& it=*ait;
```
Comment (peterschrammel) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: `const auto &it` (more instances)

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 5]
```diff
@@ -1566,18 +1873,85 @@ codet java_bytecode_convert_methodt::convert_instructions(
...
     const unsigned address=it.first;
     assert(it.first==it.second.source->address);
     const codet &c=it.second.code;
 
-    if(targets.find(address)!=targets.end())
-      code.add(code_labelt(label(std::to_string(address)), c));
-    else if(c.get_statement()!=ID_skip)
-      code.add(c);
+    if(!start_new_block)
```
Comment (peterschrammel) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Add a comment to explain what these two cases mean

**Comment by smowton** [comment]
Pushed further style fixes and rebased. Tests to come shortly...

**Comment by smowton** [comment]
Pushed a basic test -- also fixed an unrelated test failure due to mishandling class files without a variable table.

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 6]
```diff
@@ -1557,18 +1868,91 @@ codet java_bytecode_convert_methodt::convert_instructions(
...
-  for(const auto &address_pair : address_map)
+  // Try to recover block structure as indicated in the local variable table:
+
+  // The block tree node mirrors the block structure of root_block,
+  // indexing the Java PCs were each subblock starts and ends.
+  block_tree_nodet root;
+  code_blockt root_block;
+
+  bool start_new_block=true;
+  for(auto ait=address_map.begin(), aend=address_map.end(); ait!=aend; ++ait)
```
Comment (peterschrammel) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: What is the reason for not using a ranged for here?

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 7]
```diff
@@ -496,6 +550,280 @@ static member_exprt to_member(const exprt &pointer, const exprt &fieldref)
...
+          an old_label that should be replaced throughout by new_label.
+
+ Outputs: None (side-effects on repl)
+
+ Purpose: Find all goto statements in 'repl' that target 'old_label'
+          and redirect them to 'new_label'.
+
+\*******************************************************************/
+
+void replace_goto_target(
```
Comment (peterschrammel) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Please add this utility function as a static method to the class.

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 8]
```diff
@@ -1557,18 +1868,91 @@ codet java_bytecode_convert_methodt::convert_instructions(
...
+      get_or_create_block_for_pcrange(
+        root,
+        root_block,
+        v.start_pc,
+        v.start_pc+v.length,
+        std::numeric_limits<unsigned>::max(),
+        address_map);
+    }
+  }
+  for(const auto &vlist : variables)
```
Comment (peterschrammel) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Why do you need two passes here?

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 9]
```diff
@@ -325,6 +368,17 @@ void java_bytecode_convert_methodt::convert(
     variables[v.index][number_index_entries].symbol_expr=result;
     variables[v.index][number_index_entries].start_pc=v.start_pc;
     variables[v.index][number_index_entries].length=v.length;
+    symbolt new_symbol;
+    new_symbol.name=identifier;
+    new_symbol.type=t;
+    new_symbol.base_name=v.name;
+    new_symbol.pretty_name=id2string(identifier).substr(6, std::string::npos);
```
Comment (peterschrammel) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Maybe add a comment that this removes the java:: prefix

**Comment by peterschrammel** [comment]
Also see comments by @mgudemann in https://github.com/diffblue/cbmc-testgen/pull/35

File: src/java_bytecode/java_bytecode_convert_method.cpp (hunk 8 above)
Comment (smowton) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: See comment in first pass (elaborated slightly in amendment)

**Comment by smowton** [comment]
All comments addressed (including Matthias') except where noted

**Pull request merged**